### PR TITLE
nixos/clickhouse: add package option

### PR DESCRIPTION
**Pull request by GoldsteinE**
###### Motivation for this change
Consistent with other services and helps to work around #130101

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
Once you've made the change do you know how to test it on your machine (server or laptop) ?
if you are using flakes, you can change the input.url of the nixpkgs flake.
if you are using channels, you can probably change the channel to point to this PR.

Let me know if anything doesn't make sense.
Thank you for your contribution!

**Comment by aanderse** [comment]
Overlays weren't sufficient?

**Comment by GoldsteinE** [comment]
Overlay would probably be sufficient. I kinda dislike the idea that any usage of this module (and also `plausible`) _requires_ user to define a global overlay. IMO `.package` simplifies the default workflow here.

File: nixos/modules/services/databases/clickhouse.nix [hunk 1]
```diff
@@ -13,6 +13,15 @@ with lib;
 
       enable = mkEnableOption "ClickHouse database server";
 
+      package = mkOption {
+        default = pkgs.clickhouse;
+        defaultText = "pkgs.clickhouse";
+        type = types.package;
+        description = ''
+          ClickHouse package to use
```
Comment (aanderse) [inline nixos/modules/services/databases/clickhouse.nix:21]: `description` should end in a period.

File: nixos/modules/services/databases/clickhouse.nix [hunk 2]
```diff
@@ -13,6 +13,15 @@ with lib;
 
       enable = mkEnableOption "ClickHouse database server";
 
+      package = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/databases/clickhouse.nix:16]: You need to reference this option now.

**Review by aanderse — CHANGES_REQUESTED** [review]

**Comment by Ma27** [comment]
A `package` option only makes sense if you **can't** use an overlay because this would imply tons of (unneeded) rebuilds (`systemd` or `sudo` are examples here) or if you need to specify a certain version for the module (e.g. postgres, nextcloud). None of that applies here, so I'd rather close this.

**Comment by nh2** [comment]
I'm a big fan of having a `package` option for most modules:

Often in production you need to use a different version/override of the package than in other parts of the system (e.g. as a dev dependency). Often you also want to run different machines with different package versions, e.g. for rolling upgrades of distributed systems. That gets messy and error-prone quickly with conditionally applied overlays.

For this reason, we do this with many service modules, none of which are like `systemd` or `sudo`:

* `services.postgresql.package`
* `services.mysql.package`
* `services.redis.package`
* `services.consul.package`

Clickhouse should be equally configurable.

**Comment by aanderse** [comment]
`package` options are fine... but I notice that because so many `package` options exist many people don't understand what overlays are, or even know that they exist. Oh well :man_shrugging:

**Comment by nh2** [comment]
I've taken the liberty to force-push the addressing of feedback into this PR.

**Comment by happysalada** [comment]
Personally this looks good to me.

**Comment by nh2** [comment]
I have built `-A clickhouse.tests`. It passed.

File: nixos/modules/services/databases/clickhouse.nix (hunk 2 above)
Comment (nh2) [reply nixos/modules/services/databases/clickhouse.nix:16]: Done.

File: nixos/modules/services/databases/clickhouse.nix (hunk 1 above)
Comment (nh2) [reply nixos/modules/services/databases/clickhouse.nix]: Done.

File: nixos/modules/services/databases/clickhouse.nix [hunk 3]
```diff
@@ -13,6 +13,15 @@ with lib;
...
+        default = pkgs.clickhouse;
```
[the remark this answers is not in the document]
Comment (nh2) [reply nixos/modules/services/databases/clickhouse.nix:18]: Done.

**Pull request merged**